README.txt
CNNRFI.py - Main script used to train the CNN, requires access to a larger dataset for incorporating more baseline types.

FindHypers.py - Extension of CNNRFI.py that uses the same CNN structure to iterate over the parameter space and find the optimal parameters.

NeuralAssignPredict.py - Takes the output model (DEEPcnn.txt) and applies it to the rest of the dataset. Usage is 'python NeuralAssignPredict.py zen.2457691.54970.xx.HH.uvc', output file will be 'zen.2457691.54970.xx.HH.uvcr'. To view output information however requires the installation of AIPY at https://github.com/AaronParsons/aipy. To view output RFI flagged file then run 'plot_uv.py zen.2457691.54970.xx.HH.uvcr -a 9_31', which will show the log amplitude of baseline with antennas 9 and 31.


zen.2457691.54970.xx.HH.uvcA/history
CORR-DACQ: created file.
/home/obs/src/RTP/scripts/hera/correct_hsa19.py zen.2457691.54970.xx.uv
PULL ANTPOLS:/home/obs/src/RTP/scripts/hera/pull_antpols_simple.py -p xx -a 9,10,20,22,31,43,53,64,65,72,80,81,88,89,96,97,104,105,112 zen.2457691.54970.xx.uvc
PULL ANTPOLS:/users/jkerriga/capo/dcj/scripts/pull_antpols.py -a 9_10,31_105 zen.2457691.54970.xx.HH.uvc


xrfi.py
'''Module for all things Radio Frequency Interference Flagging'''
import numpy as np
from scipy.signal import medfilt

def medmin(d):
    #return np.median(np.min(chisq,axis=0))
    mn = np.min(d,axis=0)
    return 2*np.median(mn) - np.min(mn)

def medminfilt(d, K=8):
    d_sm = np.empty_like(d)
    for i in xrange(d.shape[0]):
        for j in xrange(d.shape[1]):
            i0,j0 = max(0,i-K), max(0,j-K)
            i1,j1 = min(d.shape[0], i+K), min(d.shape[1], j+K)
            d_sm[i,j] = medmin(d[i0:i1,j0:j1])
    return d_sm

#def omni_chisq_to_flags(chisq, K=8, sigma=6, sigl=2):
#    '''Returns a mask of RFI given omnical's chisq statistic'''
#    if False:
#        w_sm = np.empty_like(chisq)
#        sig = np.empty_like(chisq)
#        #get smooth component of chisq
#        for i in xrange(chisq.shape[0]):
#            for j in xrange(chisq.shape[1]):
#                i0,j0 = max(0,i-K), max(0,j-K)
#                i1,j1 = min(chisq.shape[0], i+K), min(chisq.shape[1], j+K)
#                #w_sm[i,j] = np.median(chisq[i0:i1,j0:j1])
#                w_sm[i,j] = medmin(chisq[i0:i1,j0:j1])
#    else: w_sm = medfilt(chisq, 2*K+1)
#    #the residual from smooth component
#    w_rs = chisq - w_sm 
#    w_sq = np.abs(w_rs)**2
#    #get the standard deviation of the media.
#    if False:
#        for i in xrange(chisq.shape[0]):
#            for j in xrange(chisq.shape[1]):
#                i0,j0 = max(0,i-K), max(0,j-K)
#                i1,j1 = min(chisq.shape[0], i+K), min(chisq.shape[1], j+K)
#                #sig[i,j] = np.sqrt(np.median(w_sq[i0:i1,j0:j1]))
#                sig[i,j] = np.sqrt(medmin(w_sq[i0:i1,j0:j1]))
#    else: sig = np.sqrt(medfilt(w_sq, 2*K+1))
#    #Number of sigma above the residual unsmooth part is.
#    f1 = w_rs / sig
#    return watershed_flag(f1, sig_init=sigma, sig_adj=sigl)


def watershed_flag(d, f=None, sig_init=6, sig_adj=2):
    '''Returns a watershed flagging of an array that is in units of standard
    deviation (i.e. how many sigma the datapoint is from the center).'''
    #mask off any points above 'sig' sigma and nan's.
    f1 = np.ma.array(d, mask=np.where(d > sig_init,1,0)) 
    f1.mask |= np.isnan(f1)
    if not f is None: f1.mask |= f
    
    # Loop over flagged points and examine adjacent points to see if they exceed sig_adj
    #Start the watershed
    prevx,prevy = 0,0
    x,y = np.where(f1.mask)
    while x.size != prevx and y.size != prevy:
        for dx,dy in [(1,0),(-1,0),(0,1),(0,-1)]:
            prevx,prevy = x.size, y.size
            xp, yp = (x+dx).clip(0,f1.shape[0]-1), (y+dy).clip(0,f1.shape[1]-1)
            i = np.where(f1[xp,yp] > sig_adj)[0] # if sigma > 'sigl'
            f1.mask[xp[i],yp[i]] = 1
            x,y = np.where(f1.mask)
    return f1.mask
    
def toss_times_freqs(mask, sig_t=6, sig_f=6):
    f1ch = np.average(f1.mask, axis=0); f1ch.shape = (1,-1)
    #The cut off value is a made up number here...sig = 'sig' if none flagged.
    f1.mask = np.logical_or(f1.mask, np.where(f1 > sig_init*(1-f1ch), 1, 0))
    f1t = np.average(f1.mask, axis=1) # band-avg flag vs t
    ts = np.where(f1t > 2*np.median(f1t))
    f1.mask[ts] = 1
    f1f_sum = np.sum(f1.filled(0), axis=0)
    f1f_wgt = np.sum(np.logical_not(f1.mask), axis=0)
    f1f = f1f_sum / f1f_wgt.clip(1,np.Inf)
    fs = np.where(f1f > 2)
    f1.mask[:,fs] = 1
    mask = f1.mask
    return mask

def xrfi_simple(d, f=None, nsig_df=6, nsig_dt=6, nsig_all=0):
    if f is None: f = np.zeros(d.shape, dtype=np.bool)
    if nsig_df > 0:
        d_df = d[:,1:-1] - .5 * (d[:,:-2] + d[:,2:])
        d_df2 = np.abs(d_df)**2
        sig2 = np.median(d_df2, axis=1) # XXX 1 or 0 here?
        sig2.shape = (-1,1)
        f[:,0] = 1; f[:,-1] = 1
        f[:,1:-1] = np.where(d_df2 / sig2 > nsig_df**2, 1, f[:,1:-1])
    if nsig_dt > 0:
        d_dt = d[1:-1,:] - .5 * (d[:-2,:] + d[2:,:])
        d_dt2 = np.abs(d_dt)**2
        sig2 = np.median(d_dt2, axis=0) # XXX 0 or 1 here?
        sig2.shape = (1,-1)
        f[0,:] = 1; f[-1,:] = 1
        f[1:-1,:] = np.where(d_dt2 / sig2 > nsig_dt**2, 1, f[1:-1,:])
    if nsig_all > 0:
        ad = np.abs(d)
        med = np.median(ad)
        sig = np.sqrt(np.median(np.abs(ad-med)**2))
        f = np.where(ad > med + nsig_all * sig, 1, f)
    return f

def detrend_deriv(d, dt=True, df=True):
    '''XXX This only works ok on sparse RFI.'''
    if df:
        d_df = np.empty_like(d)
        d_df[:,1:-1] = (d[:,1:-1] - .5 * (d[:,:-2] + d[:,2:])) / np.sqrt(1.5)
        d_df[:,0] = (d[:,0] - d[:,1]) / np.sqrt(2)
        d_df[:,-1] = (d[:,-1] - d[:,-2]) / np.sqrt(2)
    else: d_df = d
    if dt:
        d_dt = np.empty_like(d_df)
        d_dt[1:-1] = (d_df[1:-1] - .5 * (d_df[:-2] + d_df[2:])) / np.sqrt(1.5)
        d_dt[0] = (d_df[0] - d_df[1]) / np.sqrt(2)
        d_dt[-1] = (d_df[-1] - d_df[-2]) / np.sqrt(2)
    else: d_d= d_df
    d2 = np.abs(d_dt)**2
    # model sig as separable function of 2 axes
    sig_f = np.median(d2, axis=0); sig_f.shape = (1,-1)
    sig_t = np.median(d2, axis=1); sig_t.shape = (-1,1)
    sig = np.sqrt(sig_f * sig_t / np.median(sig_t))
    return d_dt / sig

def detrend_medminfilt(d, K=8):
    d_sm = medminfilt(np.abs(d), 2*K+1)
    d_rs = d - d_sm
    d_sq = np.abs(d_rs)**2
    sig = np.sqrt(medminfilt(d_sq, 2*K+1)) * (K/.64) # puts minmed on same scale as average
    f = d_rs / sig
    return f

def detrend_medfilt(d, K=8):
    d = np.concatenate([d[K-1::-1],d,d[:-K-1:-1]], axis=0)
    d = np.concatenate([d[:,K-1::-1],d,d[:,:-K-1:-1]], axis=1)
    d_sm = medfilt(d, 2*K+1)
    d_rs = d - d_sm 
    d_sq = np.abs(d_rs)**2
    sig = np.sqrt(medfilt(d_sq, 2*K+1) / .456) # puts median on same scale as average
    f = d_rs / sig
    return f[K:-K,K:-K]

def xrfi(d, f=None, K=8, sig_init=6, sig_adj=2):
    nsig = detrend_medfilt(d, K=K)
    f = watershed_flag(np.abs(nsig), f=f, sig_init=sig_init, sig_adj=sig_adj)
    return f

# XXX split off median filter as one type of flagger



basicNN.py
import torch
from torch.autograd import Variable
from glob import glob
import aipy as a
import numpy as n
import pylab as pl
"""
A fully-connected ReLU network with one hidden layer, trained to predict y from x
by minimizing squared Euclidean distance.
This implementation defines the model as a custom Module subclass. Whenever you
want a model more complex than a simple sequence of existing Modules you will
need to define your model this way.
"""
def loadAipyData():
    HERAlist = glob('/Users/josh/Desktop/HERA/data/*A')
    HERAdata = []
    times = []
    for l in ['9_10']: #,'31_105','10_105','9_31','9_105','10_31']:                                                                                    
        data = []
        for k in HERAlist:
            uvHERA = a.miriad.UV(k)
            a.scripting.uv_selector(uvHERA, l, 'xx')
            for p,d,f in uvHERA.all(raw=True):
                data.append(d)
                times.append(uvHERA['lst'])
        if l == '9_10':
            HERAdata = [data]
        else:
            HERAdata.append(data)
    print n.shape(HERAdata)
    HERAdata = n.array(HERAdata)
    times = n.array(times)
    return HERAdata,times

def loadAipyDataTest():
    HERAlist = glob('/Users/josh/Desktop/HERA/data/*U')
    HERAdata = []
    times = []
    for l in ['9_31']: 
#,'31_105','10_105','9_31','9_105','10_31']:                                                                                    
        data = []
        for k in HERAlist:
            uvHERA = a.miriad.UV(k)
            a.scripting.uv_selector(uvHERA, l, 'xx')
            for p,d,f in uvHERA.all(raw=True):
                data.append(d)
                times.append(uvHERA['lst'])
        if l == '9_31':
            HERAdata = [data]
        else:
            HERAdata.append(data)
    print n.shape(HERAdata)
    HERAdata = n.array(HERAdata)
    times = n.array(times)
    return HERAdata,times

def expandMask(mask,batch):
    sh = n.shape(mask)
#    expData = data
    expMask = mask
    for i in range(batch):
#        expData = n.vstack((expData,data+0.1*n.random.randn()*(n.random.randn(sh[0],sh[1])+1j*n.random.randn(sh[0],sh[1]))))
        expMask = n.vstack((expMask,mask))
    return expMask

data,time = loadAipyData()
mask = n.loadtxt('trainMask_HQ.txt')
data = n.nan_to_num(data.reshape(-1,1024))
mask = mask

#mask = expandMask(mask,5)
#print n.shape(data)
#print n.shape(mask)

class TwoLayerNet(torch.nn.Module):
  def __init__(self, D_in, H1,H2,D_out):
    """
    In the constructor we instantiate two nn.Linear modules and assign them as
    member variables.
    """
    super(TwoLayerNet, self).__init__()
    self.linear1 = torch.nn.Linear(D_in, H1)
    self.linear2 = torch.nn.Linear(H1, H2)
    self.linear3 = torch.nn.Linear(H2, D_out)
    #maybe add a dropout layer? see what that does
  def forward(self, x):
    """
    In the forward function we accept a Variable of input data and we must return
    a Variable of output data. We can use Modules defined in the constructor as
    well as arbitrary operators on Variables.
    """
    drop = torch.nn.Dropout(p=0.9)
    m = torch.nn.ReLU()
    p = torch.nn.ReLU()
    h1_relu = self.linear1(x)
    h2_relu = self.linear2(p(drop(h1_relu)))
    y_pred = self.linear3(m(h2_relu))
    return y_pred


# N is batch size; D_in is input dimension;
# H is hidden dimension; D_out is output dimension.
N, D_in, H1,H2, D_out = 390,1024,1024,1024,1024
epochs = 100
from torch.utils.data import DataLoader

# Create random Tensors to hold inputs and outputs, and wrap them in Variables
from torch.utils.data import TensorDataset
data = n.abs(data).astype(float)
data = data/n.max(n.abs(data))
#print n.shape(data)
data1 = torch.from_numpy(data)
mask1 = torch.from_numpy(mask)
Container = TensorDataset(data1,mask1)

#train_data = DataLoader(Container,batch_size=N,shuffle=True)
#x = Variable(torch.Tensor(n.abs(data[0:1000,:])))
#y = Variable(torch.Tensor(mask[0:1000,:].astype(int)), requires_grad=False).view(-1)

# Construct our model by instantiating the class defined above
model = TwoLayerNet(D_in, H1,H2,D_out)

# Construct our loss function and an Optimizer. The call to model.parameters()
# in the SGD constructor will contain the learnable parameters of the two
# nn.Linear modules which are members of the model.
criterion = torch.nn.MSELoss()
#criterion = torch.nn.CrossEntropyLoss()
optimizer = torch.optim.Adagrad(model.parameters(), lr=1e-2)
#for t in range(1000):
lossVal = 10000000.
t = 0
lossArray = []
accuracy = []
#while lossVal>800.: 
for t in range(epochs):
    train_data = DataLoader(Container,batch_size=N,shuffle=True)
    for i,j in enumerate(train_data):
  # Forward pass: Compute predicted y by passing x to the model
        x = Variable(j[0].float())
        y = Variable(j[1].float())
#        print x.mean()
#        print y.mean()
        y_pred = model(x)

#        y_pred = y_pred.view(-1,2)
        y_pred = y_pred.float()
        y = y.view(-1,1024)
#        print y_pred.size(),y.size()
  #print y_pred.data.numpy()
  # Compute and print loss
        loss = criterion(y_pred, y)
        ypred = n.array(n.round(y_pred.data.numpy()))
        ynump = n.array(n.round(y.data.numpy()))
        shp = n.shape(ynump)

        print (n.sum(ynump==ypred)).astype(float)/(shp[0]*shp[1])
        acc = (n.sum(ynump==ypred)).astype(float)/(shp[0]*shp[1])
#  print acc,'%'
#  accuracy.append(acc)
#  loss = log_loss(y_pred,y)
        print(t, loss.data[0])
        accuracy.append(acc)
  # Zero gradients, perform a backward pass, and update the weights.
        optimizer.zero_grad()
        loss.backward()
        optimizer.step()
    lossVal = loss.data[0]
    lossArray.append(lossVal)
#y_pred = y_pred[:,0]
#pred = y_pred[:,1].contiguous()
pl.subplot(211)
pl.plot(n.log10(n.array(lossArray)))
pl.subplot(212)
pl.plot(n.array(accuracy))
pl.show()

pl.subplot(211)
pl.imshow(n.abs(n.round(mask)),aspect='auto')
pl.subplot(212)
pl.imshow(n.abs(n.round(y_pred.view(-1,1024).data.numpy())),aspect='auto')
pl.show()


data_test,times_test = loadAipyDataTest()
data_test = data_test/n.abs(data_test).max()
x1 = Variable(torch.Tensor(n.abs(data_test[0,:,:])))
torch.save(model,'heranet.txt')
mask_pred = model(x1)
def delayspectrum(dat):
    DATA = n.fft.fft(dat,axis=1)
    DATA_ = n.fft.fftshift(DATA,axes=1)
    return DATA_
rfiMasked = n.round(mask_pred.data.numpy())
dataMask = data_test[0,:,:]*rfiMasked
bh = a.dsp.gen_window(1024,window='blackman-harris')
pl.subplot(211)
pl.imshow(n.log10(n.abs(dataMask)),aspect='auto',cmap='jet')
pl.subplot(212)
pl.imshow(n.log10(n.abs(delayspectrum(bh*dataMask))),aspect='auto',cmap='jet')
pl.show()


trainNN.py
import torch 
import torch.nn as nn
import torchvision.datasets as dsets
import torchvision.transforms as transforms
from torch.autograd import Variable
import aipy as a
import pylab as pl
from glob import glob
import numpy as n

# Hyper Parameters
num_epochs = 5
batch_size = 100
learning_rate = 0.001

def loadAipyData():
    HERAlist = glob('/Users/josh/Desktop/HERA/data/*A')
    HERAdata = []
    times = []
    for l in ['9_10']: #,'31_105','10_105','9_31','9_105','10_31']:                                                                                    
        data = []
        for k in HERAlist:
            uvHERA = a.miriad.UV(k)
            a.scripting.uv_selector(uvHERA, l, 'xx')
            for p,d,f in uvHERA.all(raw=True):
                data.append(d)
                times.append(uvHERA['lst'])
        if l == '9_10':
            HERAdata = [data]
        else:
            HERAdata.append(data)
    print n.shape(HERAdata)
    HERAdata = n.array(HERAdata)
    times = n.array(times)
    return HERAdata,times


# MNIST Dataset
train_dataset = dsets.MNIST(root='../data/',
                            train=True, 
                            transform=transforms.ToTensor(),
                            download=True)

test_dataset = dsets.MNIST(root='../data/',
                           train=False, 
                           transform=transforms.ToTensor())

# Data Loader (Input Pipeline)
train_loader = torch.utils.data.DataLoader(dataset=train_dataset,
                                           batch_size=batch_size, 
                                           shuffle=True)

test_loader = torch.utils.data.DataLoader(dataset=test_dataset,
                                          batch_size=batch_size, 
                                          shuffle=False)

# CNN Model (1 conv layer)
class CNN(nn.Module):
    def __init__(self):
        super(CNN, self).__init__()
        self.layer1 = nn.Sequential(
            nn.Conv2d(1, 16*4, kernel_size=(193,769), padding=0),
            nn.ReLU())
            #nn.MaxPool2d(5))
            #nn.BatchNorm2d(16),
#            nn.ReLU(),
#            nn.MaxPool2d(2))
#        self.layer2 = nn.Sequential(
#            nn.Conv2d(16, 32, kernel_size=5, padding=2),
#            nn.BatchNorm2d(32),
#            nn.ReLU(),
#            nn.MaxPool2d(2))
        self.fc = nn.Linear(1024*256, 2) ##size of input images *2
        
    def forward(self, x):
        out = self.layer1(x)
        print out.size()
#        out = self.layer2(out)
        out = out.view(-1,1024)
        out = self.fc(out)
        return out
        
cnn = CNN()


# Loss and Optimizer
#criterion = nn.CrossEntropyLoss(weight=None,)
optimizer = torch.optim.Adam(cnn.parameters(), lr=learning_rate)
data,time = loadAipyData()
mask = n.loadtxt('trainMask_severalBlines.txt')[0:256,:]
data = n.nan_to_num(data.reshape(-1,1024))
data = n.array(n.abs(data[0:256,:]))
rfiFlags = Variable(torch.Tensor(mask).int())
criterion = nn.CrossEntropyLoss(weight=None,size_average=True)
# Train the Model
for epoch in range(num_epochs):
    #for i, (images, labels) in enumerate(train_loader):
   #     print images.size()
   #     print labels.size()
    vis = Variable(torch.Tensor(data.reshape(1,1,-1,1024)))
#    rfiFlags = Variable(torch.Tensor(mask))
        
        # Forward + Backward + Optimize
    optimizer.zero_grad()
    outputs = cnn(vis)
        #outputs = outputs.transpose(0,1)
        #outputs = outputs.contiguous().view(1,-1).int()
    #rfiFlags = rfiFlags.view(1,-1)
    loss = criterion(outputs, rfiFlags)
#        loss.backward()
#        optimizer.step()
        
#        if (i+1) % 100 == 0:
#            print ('Epoch [%d/%d], Iter [%d/%d] Loss: %.4f' 
#                   %(epoch+1, num_epochs, i+1, len(train_dataset)//batch_size, loss.data[0]))

# Test the Model
cnn.eval()  # Change model to 'eval' mode (BN uses moving mean/var).
correct = 0
total = 0
for images, labels in test_loader:
    images = Variable(images)
    outputs = cnn(images)
    _, predicted = torch.max(outputs.data, 1)
    total += labels.size(0)
    correct += (predicted == labels).sum()

print('Test Accuracy of the model on the 10000 test images: %d %%' % (100 * correct / total))

# Save the Trained Model
torch.save(cnn.state_dict(), 'cnn.pkl')


CNNRFI.py
import torch 
import xrfi
import torch.nn as nn
import torchvision.datasets as dsets
import torchvision.transforms as transforms
from torch.autograd import Variable
import numpy as n
import aipy as a
import pylab as pl
from glob import glob
from scipy import signal
from scipy.signal import medfilt

# Hyper Parameters
num_epochs = 1000
batch_size = 1
learning_rate = 0.000316 #10**(-4)

def loadAipyData():
    HERAlist = glob('/users/jkerriga/data/jkerriga/HERA2/lst.*.*.HH.uv')
    HERAlist = n.sort(HERAlist)
    HERAdata = []
    times = []
    for l in ['9_10']: #,'31_105','10_31']: #,'10_105','20_105']:        
                                                                      
        data = []
        for k in HERAlist:
            uvHERA = a.miriad.UV(k)
            a.scripting.uv_selector(uvHERA, l, 'xx')
            for p,d,f in uvHERA.all(raw=True):
                data.append(d)
                times.append(uvHERA['lst'])
        data = n.array(data)
        if l == '9_10':
            HERAdata = [data[0:3900,:]]
        else:
            HERAdata.append(data[0:3900,:])
    #print n.shape(HERAdata)
    HERAdata = n.array(HERAdata)
    times = n.array(times)
    return HERAdata,times

def expandMask(data,mask,batch):
    sh = n.shape(mask)
    expData = data
    expMask = mask
    for i in range(batch):
        #expData = n.vstack((expData,data+0.001*n.random.randn()*(n.random.randn(sh[0],sh[1])+ 1j*n.random.randint(-1,2)*n.random.randn(sh[0],sh[1]))))
        expMask = n.vstack((expMask,mask))
    return expData,expMask

def injectRandomRFI(data,mask,injections):
    sh = n.shape(data)
    for i in range(injections):
        if n.abs(n.random.rand())>0.5:
            ## RFI across time
            fw = n.random.randint(1,4)
            th = n.random.randint(1,900)
            fs = n.random.randint(1,sh[1]-fw)
            ts = n.random.randint(1,sh[0]-th)
            data[ts:ts+th,fs:fs+fw] = data[ts:ts+th,fs:fs+fw]+0.001*n.random.randn()*(n.random.randn(th,fw)+ 1j*n.random.randint(-1,2)*n.random.randn(th,fw))
            mask[ts:ts+th,fs:fs+fw] = 0.
        else:
            ## RFI across freq
            fw = n.random.randint(1,100)
            th = n.random.randint(1,4)
            fs = n.random.randint(1,sh[1]-fw)
            ts = n.random.randint(1,sh[0]-th)
            data[ts:ts+th,fs:fs+fw] = data[ts:ts+th,fs:fs+fw]+0.001*n.random.randn()*(n.random.randn(th,fw)+ 1j*n.random.randint(-1,2)*n.random.randn(th,fw))
            mask[ts:ts+th,fs:fs+fw] = 0.
            
    return data,mask

def corrPass(data1,data2):
    cCout = n.zeros_like(data1)
    for i in range(n.shape(data1)[1]):
        a = data1[:,i]
        b = data2[:,i]
        cCout[:,i] = signal.correlate(a,b,mode='same')
    return cCout

def ObsMedian(data):
    sh = n.shape(data)
    for i in range(sh[0]/50):
        data[i*50:(i+1)*50] = (data[i*50:(i+1)*50]-n.median(data[i*50:(i+1)*50]))/n.std(data[i*50:(i+1)*50])
    return data

def NormByObs(data):
    sh = n.shape(data)
    for i in range(sh[0]/50):
        data[i*50:(i+1)*50] = data[i*50:(i+1)*50]/n.max(data[i*50:(i+1)*50])
    return data


data,times = loadAipyData()
data = data.reshape(-1,1024)
print n.shape(data)
XRFImask = xrfi.xrfi_simple(data,nsig_df=20,nsig_dt=20,nsig_all=25)
#data = data[0:3900,:]

mask = n.loadtxt('trainMask_HQ.txt')[0:3900,:] #n.logical_not(XRFImask[0:3900,:]).astype(int)*n.loadtxt('trainMask_HQ.txt')[0:3900,:]
data = n.nan_to_num(data.reshape(-1,1024))
from torch.utils.data import DataLoader
sh = n.shape(data)
print n.shape(data),n.shape(mask)
# Create random Tensors to hold inputs and outputs, and wrap them in Variables  
#data,mask = expandMask(data,mask,2)                                       
from torch.utils.data import TensorDataset
#data = n.vstack((data,expDat))
#mask = n.vstack((mask,expMask))
#data,mask = injectRandomRFI(data,mask,400)

data = n.abs(data)
data = NormByObs(data)
#data = (data-n.median(data))/n.std(data)
#data = n.abs(ObsMedian(data))
pl.subplot(211)
pl.imshow(n.log10(data),aspect='auto',interpolation='none')
pl.subplot(212)
pl.imshow(n.log10(data*mask),aspect='auto',interpolation='none')
pl.show()
#data = n.array([n.abs(data),n.angle(data)])
#mask = n.array([mask,mask])
mask = mask.reshape(1*78,-1,1024)
data = data.reshape(1*78,-1,1024)
#print n.shape(data)
#data = (n.abs(data))
data1 = torch.from_numpy(data)
mask1 = torch.from_numpy(mask)
train_dataset = TensorDataset(data1,mask1)
train_loader = DataLoader(train_dataset,batch_size=batch_size,shuffle=True)
test_loader = DataLoader(train_dataset,batch_size=batch_size,shuffle=False)
del(mask1)
del(data1)
del(data)
# CNN Model (2 conv layer)
class CNN(nn.Module):
    def __init__(self,dropRate):
        self.dropRate = dropRate
        super(CNN, self).__init__()

        self.layer1 = nn.Sequential(
            nn.BatchNorm2d(1),
            nn.Conv2d(1, 16*1, kernel_size=5, padding=2),
            nn.BatchNorm2d(16*1),
            nn.MaxPool2d(kernel_size=(1,16)),
            #nn.Dropout(p=self.dropRate),
            nn.ReLU())

        self.layer2 = nn.Sequential(
            nn.Conv2d(1*16, 2*16, kernel_size=5, padding=2),
            nn.BatchNorm2d(2*16),
            #nn.Dropout2d(p=self.dropRate),
            nn.MaxPool2d(kernel_size=(2,1)),
            nn.ReLU())

        self.layer3 = nn.Sequential(
            nn.Conv2d(2*16, 2*32, kernel_size=5, padding=2),
            nn.BatchNorm2d(2*32),
            #nn.MaxPool2d(kernel_size=(1,2)),
            nn.ReLU())

        self.layer4 = nn.Sequential(
            nn.Conv2d(2*32, 2*16, kernel_size=5, padding=2),
            #nn.Dropout2d(p=self.dropRate),
            nn.BatchNorm2d(2*16),
            nn.ReLU())

        self.layer5 = nn.Sequential(
            nn.Conv2d(2*8*2, 2*8*2, kernel_size=5, padding=2),
            nn.BatchNorm2d(2*8*2),
            nn.ReLU())
        
        self.layer6 = nn.Sequential(
            nn.Conv2d(2*16, 2*16, kernel_size=5, padding=2),
            nn.BatchNorm2d(2*16),
            nn.Dropout2d(p=self.dropRate),
            nn.ReLU())
        
        self.layer6_5 = nn.Sequential(
            nn.Linear(1024,1024),
            nn.ReLU())

        self.layer7 = nn.Sequential(
            nn.Linear(1024,1024),
            nn.ReLU())

        self.layer8 = nn.Sequential(
            nn.Linear(1024,1024),
            nn.ReLU())
        
        self.fc = nn.Linear(1024, 2*1024)

    def forward(self, x):
        out = self.layer1(x).view(1,16,-1,1024/16)
        out = self.layer2(out).view(1,2*16,-1,1024/16)
        out = self.layer3(out).view(1,2*32,-1,1024/16)
        out = self.layer4(out).view(1,2*16,-1,1024/16)
        out = self.layer5(out).view(1,2*16,-1,1024/16)
        out = self.layer6(out).view(-1,1024).float()
        out = self.layer6_5(out).view(-1,1024).float()
        out = self.layer7(out).view(-1,1024).float()
        out = self.layer8(out).view(-1,1024).float()
        
        out = self.fc(out)
        return out
        
cnn = CNN(0.8)

# Loss and Optimizer
criterion = nn.CrossEntropyLoss()
optimizer = torch.optim.Adam(cnn.parameters(), lr=learning_rate)
loss_array = []
correct = []

# Train the Model
for epoch in range(num_epochs):
    runAvg = 0.
    for i, (images, labels) in enumerate(train_loader):
        images = Variable(images).float()
        labels = Variable(labels).long().view(-1)
        print images.size(),labels.size()
        # Forward + Backward + Optimize
        optimizer.zero_grad()
        images = images.view(1,1,-1,1024)
        outputs = cnn(images)
        outputs = outputs.view(-1,2)
        print outputs.size()
        loss = criterion(outputs.float(), labels.long())
        loss.backward()
        optimizer.step()
        runAvg = n.mean((runAvg,loss.data[0]))
        loss_array.append(loss.data[0])
        _, predicted = torch.max(outputs.data, 1)
        correct.append(1.0*(predicted.int()==labels.data.int()).sum()/labels.size()[0])
        #if (i+1) % 100 == 0:
        print ('Epoch [%d/%d], Iter [%d/%d] Loss: %.4f' %(epoch+1, num_epochs, i+1, len(train_dataset)/batch_size, loss.data[0]))
    #correct.append(1.0*(predicted.int()==labels.data.int()).sum()/labels.size()[0])
    #loss_array.append(runAvg)
torch.save(cnn, 'DEEPcnn.txt')
# Test the Model
cnn.eval()  # Change model to 'eval' mode (BN uses moving mean/var).
correct_test = 0
total = 0
ct_array = []
ct_totarray = []
for i,(images,labels) in enumerate(test_loader):
    images = Variable(images)
    images = images.view(1,1,-1,1024).float()
    outputs = cnn(images)
    outputs = outputs.view(-1,2)
    _, predicted = torch.max(outputs.data, 1)
    total += predicted.size(0)
    ct_totarray.append(total)
    correct_test += 1.0*(predicted.int() == labels.int()).sum()
    ct_array.append(correct_test)

print n.array(ct_array)/n.array(ct_totarray)
pl.figure()
pl.subplot(211)
pl.semilogy(loss_array)
pl.xlabel('Epoch')
pl.ylabel('Loss')
pl.subplot(212)
pl.plot(correct)
pl.ylabel('Fraction Correct')
pl.xlabel('Test Batch')
pl.plot(n.array(ct_array)/n.array(ct_totarray),'r')
pl.savefig('losscorr.png')

#print('Test Accuracy of the model on the 10000 test images: %d %%' % (100 * correct / total))
pl.figure()
predicted = predicted.view(-1,1024)
labels = labels.view(-1,1024)
pl.subplot(211)
pl.imshow(n.abs(labels.numpy()),aspect='auto')
pl.subplot(212)
pl.imshow(n.abs(n.round(predicted.numpy())),aspect='auto')
pl.show()
images = images.view(-1,1024).data.numpy()
pl.subplot(211)
pl.imshow(n.log10(n.abs(images*n.round(predicted.numpy()))),aspect='auto',cmap='jet')
pl.subplot(212)
pl.imshow(n.log10(n.abs(images)),aspect='auto',cmap='jet')
pl.show()
# Save the Trained Model
#torch.save(cnn, 'DEEPcnn.txt')
